[doc] Iterators: Create new top-level README with motivation and background. (#569)
iree-org/iree-llvm-sandbox: experimental/iterators/README.md

# Iterators dialect for MLIR

This folder contains an MLIR dialect based on the concepts of database-style

> itertors, which allow to express computations on streams of data.

Reviewer: ```suggestion
iterators, which allow to express computations on streams of data.
```

(or "queries") and, increasingly so, use JIT-compilers to reduce runtime
overheads. While this seems to make MLIR a good framework for implementing the
compilers of these systems, computations of their domains have some properties
that MLIR is currently unprepared for. This dialect is a first step to bridge
that gap: its goal is to implement database-style "iterators," which database
systems typically use in their execution layer.

## Background

### Analogy to Iterators in Python

Python's concept of [iterators](https://docs.python.org/3/glossary.html#term-iterator)
is very similar to that of the database-style iterators that this project is
targeting. The subsequent explains Python iterators in order to establish
fundamental concepts and terminology in a context that is familiar to many
readers.

Python's iterators allow to work with sequences of values without materializing
these sequences. This allows to (1) save memory for storing the entire sequence,
which may allow to do computations that would otherwise exceed the main memory,
(2) avoid computing values in the sequence that are never used, (3) as a

> concequence, work with inifite sequences, and (4) improve cache efficiency

Reviewer: ```suggestion
consequence, work with infinite sequences, and (4) improve cache efficiency
```

passed between them is constant, the above example essentially works on
arbitrarily large files.

### Iterators in Database Systems

Database systems (both relational and non-relational) typically express their
computations as transformations of streams of data in order to reduce data
movement between slow and fast memory (e.g., between DRAM and CPU cache).
Transformations are typically made composable through an "iterator" interface:
Each iterator produces a stream of data that can be consumed "downstream" one
element at a time, and, in turn, consumes the streams from its zero or more
"upstream" iterators to do so -- without knowing anything about its upstream and
downstream iterators. Each iterator may have complex control flow logic, often
depending on the data that it consumes, and can manage its own state. The
iterator interface, thus, passes data *and* control: when one iterator asks its
upstream iterator for the next element, it passes control to that iterator,
which returns the control together with the next element. By limiting the number
of in-flight elements to essentially one, this minimizes the overall state and
thus data movement.

> TODO(ingomueller): explain Volcano iterator interface

Reviewer: Maybe also something about how inlining producer functions can help execution time, as this currently feels like way too many calls everywhere to be very efficient, IMO.

Maybe also something about batch-based iterators, i.e., bundling together say a cache line of data before returning.

objects. Since the state in each `iterator` object and the number of values
passed between them is constant, the above example essentially works on
arbitrarily large files.

#### Distinction to Related Concepts in Python

A few related but distinct concepts deserve a short discussion:

* `iterable`: As mentioned before, `iterable`s in Python are things than can be
  iterated over (such as containers like `list`s, `tuple`s, etc) while
  `iterator`s are objects that help with that iteration. `iterable`s provide
  `iterator`s via their `__iter__()` function. The `for` loop uses that function
  to get an `iterator`, and uses that in turn to iterate. Confusion may arise
  because `iterator`s *are* also `iterable` (in order to make `iterator`s usable
  in `for` loops).
* `generator`: This term is actually ambiguous; it may refer to
  [`generator function`](https://docs.python.org/3/glossary.html#term-generator)
  (which seems to be the more common usage of simply `generator`) or
  [`generator iterator`](https://docs.python.org/3/glossary.html#term-generator-iterator).
  In short, the former is a function that returns the latter.

>   In more detail, an `iterator function` is a function with a `yield` statement,

Reviewer 2: `iterator function` just appeared out of nowhere, did you mean `generator function` (otherwise please define the new term)

Author: Yep, a typo, it should be `generator function` -- good catch!

>   which, therefor, returns an object of the builtin type `generator` -- a class

Reviewer 2: therefore

  that implements the `iterator` interface. Similarly, a
  [`generator expression`](https://docs.python.org/3/glossary.html#term-generator-expression)
  (such as `i*i for i in range(10)`) is a language construct that, like
  `generator function`s, returns an object of the built-in type `generator`.

### Iterators in Database Systems

Database systems (both relational and non-relational) typically express their
computations as transformations of streams of data in order to reduce data
movement between slow and fast memory (e.g., between DRAM and CPU cache). They
typically do so with "iterators" that are quite similar to those in Python, and
pretty much for the same reason. First, having all transformations implement
some iterator interface makes them composable: Each iterator produces a stream
of data that can be consumed "downstream" one element at a time, and, in turn,
consumes the streams from its zero or more "upstream" iterators to do so --
without knowing anything about its upstream and downstream iterators. This helps
isolating the control flow of each iterator, which may be complex, as well as
its state. Like in Python, the iterator interface also allows for interleaved

> execution of many different computation steps py passing data *and* control:

Reviewer 2: typo: py

when one iterator asks its upstream iterator for the next element, it passes
control to that iterator, which returns the control together with the next
element. By limiting the number of in-flight elements to essentially one, this
minimizes the overall state and thus data movement.

Iterators in database systems are often implemented with some variant of the
so-called "Volcano" iterator model (named after the system that first proposed

> this model, se [Graefe'90](https://doi.org/10.1145/93605.98720)). Conceptually,

Reviewer 2: typo: se

the model is the same as those of Python's iterators; however, the communication
protocol between iterators is slightly different. Concretely, it defines that
each iterator class should have the functions `open`, `next`, and `close` (and

> is therefor sometimes called "open/next/close" iterface). Computations are

Reviewer 2: typo: therefore

work on [MonetDB/X100](https://www.cidrdb.org/cidr2005/papers/P19.pdf) for
details. (2) Alternatively, JIT compilation (of the entire computation or parts
of it) has become widespread for eliminating the function call overhead,
as popularized by [HyPer](https://doi.org/10.14778/2002938.2002940). To some
degree, the [C++ template project](examples/database-iterators-standalone)
mentioned above achieves this by relying on the C++ compiler to inline templated
function calls; another widespread technique is to produce LLVM IR one way or
the other.

### The Relational Data Model

We briefly review the data model of (relational) database systems. To a large
degree, iterators seem orthogonal to the elements they iterate over but whether
or not and how we should achieve this orthogonality may require some discussion,
so reviewing the data model here may be useful.

Relational database systems manage *relations*. A relation is a bag (i.e.,
multiset) of *tuples* (or *records*). (Textbooks often define relations as
*sets* but practical systems use on bag-based relational algebra, or even
sequence-based or mixed algebras in order to support order-sensitive operations

> such as window functions.) A record is collection of named and typed

Reviewer 2: is a collection

so reviewing the data model here may be useful.

Relational database systems manage *relations*. A relation is a bag (i.e.,
multiset) of *tuples* (or *records*). (Textbooks often define relations as
*sets* but practical systems use bag-based relational algebra, or even
sequence-based or mixed algebras in order to support order-sensitive operations
such as window functions.) A record is a collection of named and typed
*attributes* (or *fields* or, depending on the context, *columns*). Attributes
may be *nullable* (read: optional), i.e., they may contain a value of a
particular type or the special value *null* indicating the absence of a value.
In the plain relational model, attributes are atomic values (numbers, strings,
etc); in nested relational algebra, an attribute may be of a relation type or a
tuple type.

The closest equivalent to a relation in Python is arguably a
[`List[TupleType]`](https://docs.python.org/3/library/typing.html#typing.List),
where `TupleType` is some type based on
[`NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple),
e.g., `NamedTuple('Employee', [('name', Optional[str]), ('id', int)])`. One

> inaccuracy in this analogy its the fact (bag-based) relations do not specify an

Reviewer: ```suggestion
inaccuracy in this analogy its the fact that (bag-based) relations do not specify an
```
